wallet/psbt_test.go

```diff
@@ -0,0 +1,302 @@
+				return
+			}
+
+			// Check wire inputs.
+			packet := tc.packet
+			if len(packet.UnsignedTx.TxIn) != 1 {
+				t.Fatalf("expected 1 input to be added, got %d",
+					len(packet.UnsignedTx.TxIn))
+			}
+			txIn := packet.UnsignedTx.TxIn[0]
+			if txIn.PreviousOutPoint.Hash != incomingTx.TxHash() {
+				t.Fatalf("unexpected UTXO prev outpoint "+
+					"hash, got %v wanted %v",
+					txIn.PreviousOutPoint.Hash,
+					incomingTx.TxHash())
+			}
+
+			// Check partial inputs.
+			if len(packet.Inputs) != 1 {
+				t.Fatalf("expected 1 partial input to be "+
+					"added, got %d", len(packet.Inputs))
+			}
+			in := packet.Inputs[0]
+			if in.WitnessUtxo == nil {
+				t.Fatalf("partial input witness UTXO not set")
+			}
+			if !bytes.Equal(in.WitnessUtxo.PkScript, p2shAddr) {
+				t.Fatalf("unexpected witness UTXO script, "+
+					"got %x wanted %x",
+					in.WitnessUtxo.PkScript, p2shAddr)
+			}
+			if in.SighashType != txscript.SigHashAll {
+				t.Fatalf("unexpected sighash flag, got %d "+
+					"wanted %d", in.SighashType,
+					txscript.SigHashAll)
+			}
+
+			// Check outputs, find index for each of the 3 expected.
+			txOuts := packet.UnsignedTx.TxOut
+			if len(txOuts) != 3 {
+				t.Fatalf("unexpected outputs, got %d wanted 3",
+					len(txOuts))
+			}
+			p2wkhIndex := -1
+			p2wshIndex := -1
+			changeIndex := -1
+			for idx, txOut := range txOuts {
+				script := txOut.PkScript
+
+				switch {
+				case bytes.Equal(script, testScriptP2WKH):
+					p2wkhIndex = idx
+
+				case bytes.Equal(script, testScriptP2WSH):
+					p2wshIndex = idx
+
+				default:
+					changeIndex = idx
+				}
+			}
+
+			// All outputs must be found.
+			if p2wkhIndex < 0 || p2wshIndex < 0 || changeIndex < 0 {
+				t.Fatalf("not all outputs found, got indices "+
+					"p2wkh=%d, p2wsh=%d, change=%d",
+					p2wkhIndex, p2wshIndex, changeIndex)
+			}
+
+			// After BIP 69 sorting, the P2WKH output should be
+			// before the P2WSH output because the PK script is
+			// lexicographically smaller.
+			if p2wkhIndex > p2wshIndex {
+				t.Fatalf("expected output with script %x to "+
+					"be before script %x",
+					txOuts[p2wkhIndex].PkScript,
+					txOuts[p2wshIndex].PkScript)
+			}
+
+			// Finally, check the change output size and that it
+			// belongs to the wallet.
+			expectedFee := int64(368)
+			expectedChange := 1000000 - 150000 - expectedFee
+			if txOuts[changeIndex].Value != expectedChange {
+				t.Fatalf("unexpected change output size, got "+
+					"%d wanted %d",
+					txOuts[changeIndex].Value,
+					expectedChange)
+			}
+		})
+	}
+}
+
+// TestFinalizePsbt tests that a given PSBT packet can be finalized.
+func TestFinalizePsbt(t *testing.T) {
+	w, cleanup := testWallet(t)
+	defer cleanup()
+
+	// Create an address we can use to send some coins to.
+	addr, err := w.CurrentAddress(0, waddrmgr.KeyScopeBIP0084)
+	if err != nil {
+		t.Fatalf("unable to get current address: %v", addr)
+	}
+	p2shAddr, err := txscript.PayToAddrScript(addr)
```

Comment:
Should be P2WKH?  Could you add a nested P2WKH test case as well? This also applies to `FundPsbt`.

Reply:
Ah, you're right. I assigned the script to `WitnessScript` instead of `FinalScriptSig`. The added test case caught that now. Thanks!